zigpy/application.py

```diff
@@ -1155,11 +1155,21 @@ async def permit_ncp(self, time_s: int = 60) -> None:
         """
         raise NotImplementedError()  # pragma: no cover
 
-    @abc.abstractmethod
     async def permit_with_key(self, node: t.EUI64, code: bytes, time_s: int = 60):
         """Permit a node to join with the provided install code bytes."""
-        raise NotImplementedError()  # pragma: no cover
+        warnings.warn(
+            "`permit_with_key` is deprecated, use `permit_with_link_key`",
+            DeprecationWarning,
+        )
```

Comment:
Just saw this warning in the Home Assistant test logs. I think it would help to add `stacklevel=2` here, so the warning isn't emitted on `zigpy` itself but rather the location the call is coming from.